quiz/test_session.py

```diff
@@ -10,24 +11,44 @@ def test_given_new_session__has_no_correct_answers(self):
         s = UserData(session)
         self.assertSetEqual(set(), s.get_correctly_answered_questions())
 
-    def test_given_new_session__after_a_correct_answer_is_registered__it_is_listed_as_answered(self):
+    def test_given_new_session__after_a_correct_answer_to_a_published_question_is_registered__it_is_listed_as_answered(self):
+        q1 = Question.objects.create(state='PUB', difficulty=1)
+        q2 = Question.objects.create(state='PUB', difficulty=1)
         session = {}
         s = UserData(session)
-        s.register_correct_answer(23)
-        self.assertSetEqual({23}, s.get_correctly_answered_questions())
-        s.register_correct_answer(4)
-        self.assertSetEqual({4, 23}, s.get_correctly_answered_questions())
+        s.register_correct_answer(q1.pk)
+        self.assertSetEqual({q1.pk}, s.get_correctly_answered_questions())
+        s.register_correct_answer(q2.pk)
+        self.assertSetEqual({q1.pk, q2.pk}, s.get_correctly_answered_questions())
+
+    def test_given_new_session__after_a_correct_answer_to_a_retracted_question_is_registered__it_is_not_listed_as_answered(self):
+        q = Question.objects.create(state='RET', difficulty=1)
+        session = {}
+        s = UserData(session)
+        s.register_correct_answer(q.pk)
+        self.assertSetEqual(set(), s.get_correctly_answered_questions())
+
+    def test_given_new_session__after_a_published_question_becomes_retracted__its_answer_is_not_listed_as_answered(self):
+        q = Question.objects.create(state='PUB', difficulty=1)
+        session = {}
+        s = UserData(session)
+        s.register_correct_answer(q.pk)
```

Comment:
Maybe add an extra assertion here to demonstrate that first it's there, then it's not?
```suggestion
        s.register_correct_answer(q.pk)
        self.assertSetEqual({q.pk}, s.get_correctly_answered_questions())
```

Reply:
Done. Also rebased.